### ISSUE TITLE
mutation with relations

### ISSUE THREAD

**Issue by AriaMoradi** [blb-ventures/strawberry-django-plus#87]
This is my code:

`models.py`
```
class User(AbstractUser):
    // some fields
    ....

class Profile(models.Model):
    user = models.OneToOneField(to=User, on_delete=models.CASCADE, primary_key=True)
    // some fields
    ...
```
`types.py`
```
@gql.django.type(models.User)
class User(gql.Node):
    username: gql.auto
    avatar: gql.auto
    profile: 'Profile'


@gql.django.partial(models.User)
class UserInputPartial(gql.NodeInput):
    username: gql.auto
    avatar: gql.auto
    profile: gql.auto

@gql.django.type(models.Profile)
class Profile(relay.Node):
    about: gql.auto
    gender: gql.auto
    header: gql.auto
    trip_status: gql.auto


@gql.django.partial(models.Profile)
class ProfileInputPartial(gql.NodeInputPartial):
    about: gql.auto
    gender: gql.auto
    header: gql.auto
    trip_status: gql.auto
```
`mutations.py` which  goes into schema
```
@gql.type
class UserMutations:
    update_user: types.User = gql.django.update_mutation(types.UserInputPartial)
```

And when I run this Mutation:
```
mutation MyMutation {
  updateUser(input: {id: "UHJvZmlsZTox"}) {
    ... on User {
      id
    }
  }
}
```
I get this kind of result
```
{
  "data": null,
  "errors": [
    {
      "message": "<class 'user.models.User'> expected, found <Profile: profile for user 'TestUser'>",
      "locations": [
        {
          "line": 3,
          "column": 3
        }
      ],
      "path": [
        "updateUser"
      ]
    }
  ]
}
```

**Comment by bellini666** [comment]
Hi @AriaMoradi ,

I'm sorry, could you provide a more detailed explanation of what your issue is?

**Comment by AriaMoradi** [comment]
I want to be able to potentially change User and it's associated profile with the same mutation.
Changing profile requires `id` while `user.profile` should have access to it without any extra steps

**Comment by mhdismail** [comment]
I gave it a try, and yes a case is not handled properly.

```
def update_m2m(
    info: Info,
    instance: Model,
    field: Union[ManyToManyField, ForeignObjectRel],
    value: Any,
):
    if value is UNSET:
        return

    if isinstance(field, OneToOneRel):
        remote_field = field.remote_field
        value, data = _parse_pk(value, remote_field.model)
        if value is None:
            value = getattr(instance, field.name)
        else:
            remote_field.save_form_data(value, instance)
            value.save()

        # If data was passed to the field, update it recursively
        if data:
            update(info, value, data)
        return
        
        ....
```

What do you think @bellini666 ? I added a check in case the value returned is None (because id wasn't passed, I think), I get the current `instance.<field>`.

After doing the above, a one to one relation mutation will work without passing the ID.

**Comment by mhdismail** [comment]
I'll also add my other point here related to mutations, given that this issue is about mutations.

> Currently if you have a reverse foreign relation and you have the following type domains: Optional[List[DomainInputPartial]], add works fine, set and remove do not work in one case. You're treating both as m2m, so this function will fail update_m2m(info, instance, field, value) if you have a foreign key that is not nullable, because .remove will not exist for the RelatedManager.
> 
> > For ForeignKey objects, this method only exists if null=True. If the related field can’t be set to None (NULL), then an object can’t be removed from a relation without being added to another. In the above example, removing e from b.entry_set() is equivalent to doing e.blog = None, and because the blog [ForeignKey](https://docs.djangoproject.com/en/dev/ref/models/fields/#django.db.models.ForeignKey) doesn’t have null=True, this is invalid.
> 
> You'd have to call a filter and delete for both the set/remove to work.

**Comment by bellini666** [comment]
Hi guys,

@mhdismail I think your solution is the way to go here! Would you like to open a PR doing that change? I'll gladly review and do a new release once it is merged :)

> I'll also add my other point here related to mutations, given that this issue is about mutations.

Yeah, I did not handle this case yet. Maybe we should check if the field can be `null` and, if not, call the related's model `.delete()` method instead of `manager.remove`.

Another option would be to add a new `delete`, so we would have `add`/`set`/`remove`/`delete`, the later being used to really delete the related object instead of removing it from the manager. Don't know if this extra field is necessary, but it would at least make it more explicit that you are really deleting the related object instead of only disassociating it from the model itself. What do you think?

**Comment by mhdismail** [comment]
I just created a PR with the change.

Regarding the `delete`, the only problem is that it might make things a bit confusing to the frontend/mobile developer(s), I say keep `remove`, the devs should not worry about how things are being handled behind the scene. I prefer checking if the field is null, this way the api is more standard regardless of this specific case.

**Comment by bellini666** [comment]
> I just created a PR with the change.

Thank you! :). I'm reviewing it and will do a new release as soon as it is merged

> Regarding the `delete`, the only problem is that it might make things a bit confusing to the frontend/mobile developer(s), I say keep `remove`, the devs should not worry about how things are being handled behind the scene. I prefer checking if the field is null, this way the api is more standard regardless of this specific case.

I agree with you! If you want to send a PR for that case as well. This one would be nice to have a test for it too, since we would be dealing with a sensitive operation, to make sure that we call remove/delete correctly depending on the situation
